https://bugs.gentoo.org/387055
http://sourceware.org/bugzilla/show_bug.cgi?id=12845

From 7ab1d000c11675aaea3b078e3e2a6cd3f1beaa08 Mon Sep 17 00:00:00 2001
From: Nick Clifton <nickc@redhat.com>
Date: Thu, 9 Jun 2011 10:36:18 +0000
Subject: [PATCH] 	PR ld/12845 	* elf.c
 (_bfd_elf_init_private_section_data): Add an assertion 
 that the output section has an allocated
 bfd_elf_section_data 	structure. 	* elfxx-mips.c
 (mips_elf_check_symbols): Do not create a stub for 
 symbols in sections that have been removed by garbage
 collection.

	* emultempl/mipself.em (mips_add_stub_section): Do not add stubs
	for sections that have been removed by garbage collection.
---
 bfd/ChangeLog           |    9 +++++++++
 bfd/elf.c               |    2 ++
 bfd/elfxx-mips.c        |    5 +++++
 ld/ChangeLog            |    6 ++++++
 ld/emultempl/mipself.em |    5 +++++
 5 files changed, 27 insertions(+), 0 deletions(-)

2011-06-09  Nick Clifton  <nickc@redhat.com>

	PR ld/12845
	* elf.c (_bfd_elf_init_private_section_data): Add an assertion
	that the output section has an allocated bfd_elf_section_data
	structure.
	* elfxx-mips.c (mips_elf_check_symbols): Do not create a stub for
	symbols in sections that have been removed by garbage collection.

diff --git a/bfd/elf.c b/bfd/elf.c
index 3d44ef8..4664c58 100644
--- a/bfd/elf.c
+++ b/bfd/elf.c
@@ -6272,6 +6272,8 @@ _bfd_elf_init_private_section_data (bfd *ibfd,
       || obfd->xvec->flavour != bfd_target_elf_flavour)
     return TRUE;
 
+  BFD_ASSERT (elf_section_data (osec) != NULL);
+
   /* For objcopy and relocatable link, don't copy the output ELF
      section type from input if the output BFD section flags have been
      set to something different.  For a final link allow some flags
diff --git a/bfd/elfxx-mips.c b/bfd/elfxx-mips.c
index 5c4636a..d5f677f 100644
--- a/bfd/elfxx-mips.c
+++ b/bfd/elfxx-mips.c
@@ -1727,6 +1727,11 @@ mips_elf_check_symbols (struct mips_elf_link_hash_entry *h, void *data)
 
   if (mips_elf_local_pic_function_p (h))
     {
+      /* PR 12845: If H is in a section that has been garbage
+	 collected it will have its output section set to *ABS*.  */
+      if (bfd_is_abs_section (h->root.root.u.def.section->output_section))
+	return TRUE;
+
       /* H is a function that might need $25 to be valid on entry.
 	 If we're creating a non-PIC relocatable object, mark H as
 	 being PIC.  If we're creating a non-relocatable object with

2011-06-09  Nick Clifton  <nickc@redhat.com>

	PR ld/12845
	* emultempl/mipself.em (mips_add_stub_section): Do not add stubs
	for sections that have been removed by garbage collection.

diff --git a/ld/emultempl/mipself.em b/ld/emultempl/mipself.em
index 7a13d4d..ada0786 100644
--- a/ld/emultempl/mipself.em
+++ b/ld/emultempl/mipself.em
@@ -138,6 +138,11 @@ mips_add_stub_section (const char *stub_sec_name, asection *input_section,
   lang_output_section_statement_type *os;
   struct hook_stub_info info;
 
+  /* PR 12845: If the input section has been garbage collected it will
+     not have its output section set to *ABS*.  */
+  if (bfd_is_abs_section (output_section))
+    return NULL;
+
   /* Create the stub file, if we haven't already.  */
   if (stub_file == NULL)
     {
-- 
1.7.6.1

